carsus/io/cmfgen/base.py

```diff
@@ -809,3 +838,116 @@ def _get_levels_lines(self, data):
+
+            label_ind_mapping = {
+                label: index for label, index in zip(levels_combine.label, levels_combine.level_index)
+            }
+            
+            label_g_mapping = {
+                label: g for label, g in zip(levels.label, levels.g)
+            }
+
+            gi, lower_level_index, upper_level_index = [], [], []
+
+            for ll, ul in zip(collisions.label_lower, collisions.label_upper):
+                if ll in label_ind_mapping:
+                    lower_level_index.append(label_ind_mapping[ll])
+                else:
+                    lower_level_index.append(np.nan)
+                
+                if ll in label_g_mapping:
+                    gi.append(label_g_mapping[ll])
+                else:
+                    gi.append(np.nan)
+
+                if ul in label_ind_mapping:
+                    upper_level_index.append(label_ind_mapping[ul])
+                else:
+                    upper_level_index.append(np.nan)
+
+            (
+                collisions["level_number_lower"],
+                collisions["level_number_upper"],
+                collisions["gi"]
+            ) = (lower_level_index, upper_level_index, gi)
+
+            collisions = collisions.dropna(subset=['level_number_lower', 'level_number_upper'])
+            
+            collisions["atomic_number"] = ion[0]
+            collisions["ion_number"] = ion[1]
+            collisions = collisions.drop(columns=["label_lower", "label_upper"])
+
+            collisions = collisions.astype({
+                'level_number_upper': int,
+                'level_number_lower': int,
+            })
+
+            collisions = collisions.set_index(
+                [
+                    "atomic_number",
+                    "ion_number",
+                    "level_number_lower",
+                    "level_number_upper",
+                ]
+            )
+            # divide the dataframe by gi and remove the column
+            collisions = collisions.iloc[:,:-1].div(collisions.gi, axis=0)
+            collisions.columns = collisions.columns.map(float)
+            
+            col_interpolator.append(interpolate.interp1d(
+                collisions.columns,
+                collisions.values,
+                bounds_error=False,
+                axis=1,
+            ))
+            
+            t_grid.extend(collisions.columns)
+            col_list.append(collisions)
+        
+        if temperature_grid is None:
+            temperature_grid = sorted(list(set(t_grid)))
+
+        col_interp = []
+        for ion_col_data, interpolator in zip(col_list, col_interpolator):
+            ion_interp = pd.DataFrame(
+                interpolator(temperature_grid),
+                index=ion_col_data.index,
+                columns=temperature_grid
+            )
+            ion_interp = ion_interp.fillna(method="bfill", axis=1)
+            ion_interp = ion_interp.fillna(method="ffill", axis=1)
```

Comment:
This fills the missing values with the last interpolated value, which is not necessarily the same as the last tabulated value. For example, if the requested temperature grid is `[2500, 7500, 12500]` and the values are tabulated at `[2000, 8000, 11000]`, the code will insert the interpolated value at 7500 K for 12500 K and not the tabulated value at 11000 K.

Reply:
Using Numpy's interp instead of Scipy's interp1d which replaces values outside the tabulated range with the last tabulated value to deal with this issue.
[cd86ea6](https://github.com/tardis-sn/carsus/pull/272/commits/cd86ea698e97f28eddacfdba23ca22773c4a6486)